### ISSUE TITLE
Beyond regex: classifier second opinion, honeypot tools, sandbox detonation, activation probes

### ISSUE THREAD

**Issue by chrisbonilla95** [UnityInFlow/injection-scanner#111]
**Where a regex scanner stops, and what the next four layers look like.** Written in answer to "what would it take to get this tool to the next level", after measuring v0.1.0 and the #110 branch against real pages, ten hand-written attacks, and the published benchmarks.

## Where the ceiling is

A lexical scanner tops out where the attack is an ordinary sentence. On the ten-attack web corpus, the branch went from 5 to 10 at HIGH by adding *shapes*, but every one of those shapes is a phrase family, and an attacker who has read `patterns/core/` writes around it in one edit. The published numbers say the same thing about every lexical and even every fine-tuned detector:

| Evidence | What it says |
|---|---|
| Lakera PINT benchmark | Best commercial API 95.2; open-weight classifiers (ProtectAI DeBERTa, Prompt Guard 2 86M) about 79 |
| Meta's AgentDojo eval on the Prompt Guard 2 card | 81% attack prevention for PG2-86M, 22% for the DeBERTa model that tops PINT next to it. Benchmark rank and agent-setting value are different things |
| *Defenses Against Prompt Attacks Learn Surface Heuristics* (2601.07185) | Fine-tuned detectors learn position, trigger tokens and topic; out-of-distribution accuracy drops up to 40% |
| *Bypassing LLM Guardrails* (2504.11168) | Character-level and adversarial evasion reach up to 100% against Azure Prompt Shield and Prompt Guard |
| *Prompt Injection Detection is Regime-Dependent* (2605.26999) | No single detector wins; transformers strongest overall, structural signals help at low FPR; thresholds decide deployment outcomes |
| LivePI (2605.17986) | On live agents, 10–30% attack success across frontier models; the defense that held was **prompt filtering plus pre-execution tool authorization**, not either alone |

So "next level" is not a better regex library, and it is not swapping the regex for an LLM. It is **composing independent signals**, each cheap to evaluate and each failing differently, in front of an action gate. The scanner already has the two things that make that possible: a measured corpus with pinned recall, and a confidence field on every finding.

### 1. Local classifier as a second opinion (weeks, not months)

**What.** An opt-in `--semantic` pass that runs a small open-weight classifier over each document's prose and emits its score as a finding-level signal. Prompt Guard 2 22M is the candidate: 22M parameters, a published ONNX export (`shisa-ai/promptguard2-onnx`, 283 MB fp32, int8 would be a quarter of that), CPU inference in tens of milliseconds per 512-token window, multilingual. `ort` (ONNX Runtime bindings) or `candle` in Rust; `parry-guard` already does exactly this for coding agents, so the plumbing is known.

**Why this first.** It is the only layer that closes the *paraphrase* gap without per-language pattern work, and on the same ten-attack corpus the DeBERTa cousin caught the Czech payload with zero language-specific effort. It composes with the regex rather than replacing it: regex is explainable and catches encoding tricks the tokenizer erases; the classifier catches phrasing the regex never saw.

**Design.**
- A non-default cargo feature (`semantic`) so the default binary stays dependency-free and 5 MB.
- Model downloaded on first use to a cache keyed by version, verified against a pinned SHA256, exactly the way spec-ci-plugin verifies this binary. Never fetched at build time.
- Output is a new context-like field, not a new pattern id: `semantic_score: 0.97` on the report, and a synthetic finding `SEM001 classifier-flagged` at MEDIUM when the score clears a threshold and *no regex fired*. Regex findings stay authoritative; the classifier only adds.
- Recall and clean-corpus gates run the classifier too, so the threshold is pinned by the same tests that pin PI-numbers. The regime-dependence paper is the argument for pinning the threshold against *this* corpus rather than trusting the model card.
- Licence: Llama 4 Community License is gated; the user accepts it when the model is fetched, the repo never vendors weights. If that's unacceptable, ProtectAI's DeBERTa v2 is Apache-2.0 but four times larger and weaker on agent payloads.

### 2. Honeypot tool definitions in the Claude Code hook (days)

**What.** The `PostToolUse` hook on the branch reports findings. The cheapest *behavioural* signal is the one the injection-lab already discovered by accident: give the agent something to do that only an injected instruction would ask for, and watch. Concretely, a `PreToolUse` hook that treats a call to a tool no legitimate task invokes — `reveal_system_prompt`, `disable_safety_checks`, `get_admin_access` — or a `Bash` command touching a planted fake credential file as a confirmed injection event. The "seven cross-domain techniques" paper (2604.18248) reports 100% precision for honeypot tool definitions by construction, and the agent-canary and Microsoft agent-governance-toolkit tutorials ship the same idea.

**Why.** Regex and classifiers judge *text*. A honeypot judges *behaviour*, which is what actually matters and what no paraphrase evades. It costs nothing until it fires.

**Design.**
- `injection-scanner canary plant` writes a fake `~/.aws/credentials`-style file with a random token and registers it; the `PreToolUse` hook exits 2 (block) on any Bash command referencing the path, and the `PostToolUse` hook flags any tool result or model output containing the token.
- Reuse `tools/injection-lab/manifest.json`'s canary machinery, and its lesson: position matters, a refusal that quotes the token is not a hit.

### 3. Sandbox detonation, the "LLM judging in a sandbox" idea done right (a quarter)

**What.** Not an LLM reading the document and voting. That design has two published failure modes: the judge is itself injectable (JudgeDeceiver, 2403.17710; targeted injections beat judges >30% of the time in 2505.13348), and the judge's precision is poor (52% precision in the TAP judge of 2606.10525). The version that works is the malware-sandbox model: **run the document through a real agent that has only canary tools, in a throwaway container, and record what it tries to do.** That is what `tools/injection-lab/agent-test.sh` already does by hand, one document at a time.

**Design.**
- `injection-scanner detonate <file>`: spawns `claude -p` (or any agent CLI) in a container with no network except the model API, a scratch filesystem, and an MCP server exposing only honeypot tools plus a canary file. The system prompt is a fixed benign task ("summarise this document"). The verdict is *which tools were called and whether the canary moved*, never the model's opinion of the text.
- Spotlighting (2606.10525's related work; Microsoft's datamarking) on the document boundary, so the agent under test is at least as defended as a production one; a hit then means the payload beats a defended agent.
- Cost: one model call per document, seconds, so this is a CI job on changed skill files and a `--detonate` flag on the pre-commit hook, not the default path.
- Output joins the same report: `DET001 agent-called-honeypot-tool` at CRITICAL with the tool call as evidence. The recall corpus gains a third column: regex, classifier, detonation.

**What it does not need.** A separate judging model. The scanner's job is to notice that the agent tried to `curl` the canary, and a shell log answers that with no LLM in the loop.

### 4. Task-drift probes, for the people who run their own models (research track)

**What.** Microsoft's TaskTracker (2406.00799) reads the model's *activations* before and after it ingests external content; a linear probe on the activation delta detects task drift at 0.99 AUC out of distribution. Attention Tracker (NAACL 2025) does it from attention on the original instruction. *Caught in the Act(ivation)* (2606.04141) applies the same idea to credential exfiltration, pre-output.

**Why it is last.** It needs white-box access, so it applies only to self-hosted open-weight agents. But for those users it is the strongest signal in the literature, and it is the only one that catches an injection the model *acted on* rather than one it *read*.

**Design.** Out of scope for the scanner binary; belongs as an `injection-scanner-probe` sidecar for vLLM/llama.cpp users, emitting the same report schema. Worth an issue and a spike, not a milestone.

## How they compose

```
document ──► regex (PI001–PI113) ───────┐
         ──► classifier (SEM, opt-in) ──┼──► one report, one confidence, one exit code
         ──► detonation (DET, CI only) ─┘
agent    ──► honeypot hook (blocks) ◄── independent of the document text
```

Each layer fails differently: regex on paraphrase, classifier on adversarial tokens, detonation on cost and non-determinism, honeypots on attacks that never touch a tool. LivePI's result is the design principle: filtering plus an action gate beat either alone. The scanner owns the filtering side and can *feed* the action gate, which is what `classifierContext` on the branch already does for Claude Code's auto-mode classifier.

## What I'd do this month

1. Land layer 2 in the hook; it is a day of Python and reuses the lab's canary code.
2. Spike layer 1 behind a cargo feature with the 22M ONNX model, and put its threshold under the recall test.
3. Turn `agent-test.sh` into `detonate` with a JSON verdict, still one document at a time.

Sources: [PINT benchmark](https://github.com/lakeraai/pint-benchmark), [Prompt Guard 2 card](https://huggingface.co/meta-llama/Llama-Prompt-Guard-2-86M), [Prompt Guard 2 ONNX](https://huggingface.co/shisa-ai/promptguard2-onnx), [parry-guard](https://github.com/vaporif/parry-guard), [Surface heuristics](https://arxiv.org/abs/2601.07185), [Bypassing guardrails](https://arxiv.org/abs/2504.11168), [Regime-dependent detection](https://arxiv.org/abs/2605.26999), [LivePI](https://arxiv.org/abs/2605.17986), [JudgeDeceiver](https://arxiv.org/abs/2403.17710), [Judge injectability](https://arxiv.org/abs/2505.13348), [Automated injection attacks](https://arxiv.org/abs/2606.10525), [Seven cross-domain techniques](https://arxiv.org/abs/2604.18248), [agent-canary](https://github.com/Thomaszhou22/agent-canary), [Microsoft agent-governance-toolkit](https://microsoft.github.io/agent-governance-toolkit/tutorials/09-prompt-injection-detection/), [TaskTracker](https://github.com/microsoft/TaskTracker), [Attention Tracker](https://github.com/khhung-906/Attention-Tracker), [Caught in the Act(ivation)](https://arxiv.org/abs/2606.04141).

---
Posted by Claude after discussing with @chrisbonilla95

**Comment by mthamil107** [comment]
On layer 2 — the honeypot tools are already built, if it saves you the day.

`d034_honeypot_tool` shipped in prompt-shield v0.7.3, Apache-2.0: `src/prompt_shield/detectors/d034_honeypot_tool.py` and `src/prompt_shield/tool_guard/_honeypot.py`, with tests and OWASP/MITRE ATLAS mappings. Two details in there that cost us time and are not obvious up front:

It needs two separate modes, not one. A model that *mentions* a decoy name while refusing is not the same event as a model that *invokes* it, and treating them alike gives you false positives on exactly the transcripts you most want to keep. That is the same wall you describe in your own lab note as "position matters, a refusal that quotes the token is not a hit" — we hit it too and ended up with text-mention and context-invocation as distinct detections.

And tool calls arrive in four different container shapes depending on the caller — flat, flat under an alternate key, OpenAI-nested, and nested under `tool` — so matching one shape silently misses the rest. That part is in the detector rather than the tool_guard helper.

One thing worth knowing if you are aiming for parity: the names in the paper are illustrative, and the shipped defaults are different — `admin_shell`, `internal_debug_console`, `system_key_dump`, `sudo_execute_raw`, `priv_esc_helper`. We moved off the paper's names deliberately, because a decoy that reads like an obvious trap gets avoided rather than triggered.

On sequencing layer 1 first, one data point rather than an objection. Against the Liu et al. task-hijacking attacks (USENIX Security 2024), a DeBERTa classifier on top of the regex firewall scored *identically* to regex alone — both catch 100% of explicit-override phrasing and about 40% of attacks where the injected instruction reads as a plausible task request. The reason it did not help is the interesting part: that classifier is trained on prompt-injection-style content, and in those attacks every individual token is, in isolation, a legitimate request, so there is no distributional signal for it to find.

Your setup differs — Prompt Guard 2 is a different training set and fine-tune from ProtectAI's model, and your corpus is not ours — so this may not transfer. But it is the specific gap the classifier is being bought to close, and one classifier already failed to close it.

If it is useful, the cheaper partial answer we found for paraphrase was sequence alignment rather than a model: Smith-Waterman, `d028`, +34.5 pp F1 on deepset at zero false-positive cost on that set. It is not free everywhere — it cost 10 additional false positives on NotInject, which we have not yet tuned out.
